python/setup.py

```diff
@@ -117,79 +124,99 @@ def get_cuda_version_from_header(cuda_include_dir):
+        language="c++",
+        extra_compile_args=["-std=c++17"],
+    ),
+    # tests:
+    Extension(
+        "*",
+        sources=cython_tests,
+        include_dirs=include_dirs,
+        library_dirs=library_dirs,
+        runtime_library_dirs=[
+            cuda_lib_dir,
+            os.path.join(os.sys.prefix, "lib"),
+        ],
+        libraries=["cuda", "cudart"],
+        language="c++",
+        extra_compile_args=["-std=c++17"],
+    ),
+]
 
 
-# cuda:
-extensions += cythonize(
-    [
-        Extension(
-            "*",
-            sources=["rmm/_cuda/*.pyx"],
-            include_dirs=include_dirs,
-            library_dirs=library_dirs,
-            runtime_library_dirs=[
-                cuda_lib_dir,
-                os.path.join(os.sys.prefix, "lib"),
-            ],
-            libraries=["cuda", "cudart"],
-            language="c++",
-            extra_compile_args=["-std=c++14"],
-        )
-    ],
-    nthreads=nthreads,
-    compiler_directives=dict(
-        profile=False, language_level=3, embedsignature=True,
-    ),
-)
+def remove_flags(compiler, *flags):
+    for flag in flags:
+        try:
+            compiler.compiler_so = list(
+                filter((flag).__ne__, compiler.compiler_so)
+            )
+        except Exception:
+            pass
+
 
-# tests:
-extensions += cythonize(
-    [
-        Extension(
-            "*",
-            sources=cython_tests,
-            include_dirs=include_dirs,
-            library_dirs=library_dirs,
-            runtime_library_dirs=[
-                cuda_lib_dir,
-                os.path.join(os.sys.prefix, "lib"),
-            ],
-            libraries=["cuda", "cudart"],
-            language="c++",
-            extra_compile_args=["-std=c++14"],
+class build_ext_no_debug(_build_ext):
+    def build_extensions(self):
+        # Full optimization
+        self.compiler.compiler_so.append("-O3")
+        # No debug symbols, full optimization, no '-Wstrict-prototypes' warning
+        remove_flags(
+            self.compiler, "-g", "-G", "-O1", "-O2", "-Wstrict-prototypes"
         )
-    ],
-    nthreads=nthreads,
-    compiler_directives=dict(
-        profile=True, language_level=3, embedsignature=True, binding=True
-    ),
-)
+        super().build_extensions()
+
+    def finalize_options(self):
+        if self.distribution.ext_modules:
+            # Delay import this to allow for Cython-less installs
+            from Cython.Build.Dependencies import cythonize
+
+            nthreads = getattr(self, "parallel", None)  # -j option in Py3.5+
+            nthreads = int(nthreads) if nthreads else None
+            self.distribution.ext_modules = cythonize(
+                self.distribution.ext_modules,
+                nthreads=nthreads,
+                force=self.force,
+                gdb_debug=False,
+                compiler_directives=dict(
+                    profile=False, language_level=3, embedsignature=True
+                ),
+            )
```

Comment:
So I think Cython already [handles this]( https://github.com/cython/cython/commit/1c4f32b93570b41de3c8028711549a32e15b31f1 ) if we use `new_build_ext`. This is what we do in [ucx-py as an example]( https://github.com/rapidsai/ucx-py/blob/b804d1a00b9a83050153612c4ab76192555dbe47/setup.py#L13 )

Reply:
@jakirkham it does, but `new_build_ext` doesn't pass the additional arguments to `cythonize` like we are here.

Reply:
One can tack on the `cython_directives` attribute to the `Extension` objects (like [done here]( https://github.com/dask/distributed/pull/4292/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R48-R53 )). Alternatively one can use the [`Extension` object from Cython to specify `cython_directives`]( https://github.com/cython/cython/blob/efdf996dced075682d86b7e53a12b56b5ebfea38/Cython/Distutils/extension.py#L35 ).

Reply:
When I build cuGraph branch-21.08 [which sets `e.cython_directives`](https://github.com/rapidsai/cugraph/blob/bf64c2c8e5cdbb4bc0d397702c335652aa31aa75/python/setup.py#L139-L142), I see this warning for every file:
```
lib/python3.8/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /home/ptaylor/dev/rapids/cugraph/python/cugraph/dask/centrality/mg_katz_centrality_wrapper.pyx
```

I also wasn't seeing these values propagate to `cythonize()` when I was debugging. Maybe this should be possible and there's a bug in cython, distutils, or how we're using it, but what I have here definitely works.